lib/asciidoctor/parser.rb

```diff
@@ -1234,7 +1235,10 @@ def self.catalog_inline_anchors text, document
     text.scan(InlineAnchorRx) do
       # lead with assignments for Ruby 1.8.7 compat
       id, reftext = $1 || $3, $2 || $4
-      document.register :ids, [id, reftext] unless $&.start_with? '\\'
+      unless $&.start_with? '\\'
+        document.register(:ids, [id, reftext])
+        document.register(:refs, [id, reftext])
```

Comment:
After reading more of the code, it feels the `Inline` object would be the right thing to save but those are only created during substitution in the conversion phase.
I would probably try to introduce a value object that gets saved into :refs. It just wraps the block/section and in the case of an inline anchor, it knows from which index of lines the ref is coming. Just so we have a pointer to the origin of the ref in all cases.